code/modules/construction/airlocks.dm

```diff
@@ -0,0 +1,138 @@
+/obj/item/frame/controller
+	name = "embedded controller frame"
+	build_machine_type = /obj/machinery/embedded_controller
+	icon = 'icons/obj/airlock_machines.dmi'
+	icon_state = "airlock_control_off"
+
+/obj/item/frame/controller/radio/
+	var/id_tag = ""
+
+/obj/item/frame/controller/radio/attackby(obj/item/C, mob/user)
+	if(istype(C,/obj/item/device/debugger))
+		var/new_tag = input(user, "Enter a new controller ID tag.", "Embedded Controller Tag Control") as null|text
+		if(new_tag)
+			id_tag = new_tag
+
+/obj/item/frame/controller/radio/airlock
+	name = "airlock controller frame"
+	build_machine_type = /obj/machinery/embedded_controller/radio/airlock
+
+/obj/item/frame/controller/radio/docking_hatch
+	name = "docking hatch frame"
+	build_machine_type = /obj/machinery/embedded_controller/radio/simple_docking_controller
+
+/obj/item/frame/controller/try_build(turf/on_wall)
+	if(!build_machine_type)
+		return
+
+	if (get_dist(on_wall,usr)>1)
+		return
+
+	var/ndir
+	if(reverse)
+		ndir = get_dir(usr,on_wall)
+	else
+		ndir = get_dir(on_wall,usr)
+
+	if (!(ndir in cardinal))
+		return
+
+	var/turf/loc = get_turf(usr)
+	if(gotwallitem(loc, ndir))
+		to_chat(usr, span("danger", "There's already an item on this wall!"))
+		return
+
+	var/obj/machinery/embedded_controller/M = new build_machine_type(loc, ndir, 1)
+	M.fingerprints = src.fingerprints
+	M.fingerprintshidden = src.fingerprintshidden
+	M.fingerprintslast = src.fingerprintslast
+	qdel(src)
+
+/obj/item/frame/controller/radio/try_build(turf/on_wall)
```

Comment:
this looks the same as the parent's? What's different?

Reply:
`var/obj/machinery/embedded_controller/M = new build_machine_type(loc, ndir, 1)`, iirc.
I couldn't figure out any way to change this without code duplication.